github_issue_summarization/distributed/train.py

```diff
@@ -14,164 +15,208 @@
+    title_pp = processor(append_indicators=True, keep_n=4500,
+      padding_maxlen=12, padding='post')
 
-  with open(data_dir + 'title_pp.dpkl', 'wb') as f:
-    dpickle.dump(title_pp, f)
+    # process the title data
+    train_title_vecs = title_pp.fit_transform(train_title_raw)
 
-  # Save the processed data
-  np.save(data_dir + 'train_title_vecs.npy', train_title_vecs)
-  np.save(data_dir + 'train_body_vecs.npy', train_body_vecs)
-else:
-  time.sleep(120)
+    print('\noriginal string:\n', train_title_raw[0])
+    print('after pre-processing:\n', train_title_vecs[0])
+
+    # Save the preprocessor
+    with open(data_dir + 'body_pp.dpkl', 'wb') as f:
+      dpickle.dump(body_pp, f)
+
+    with open(data_dir + 'title_pp.dpkl', 'wb') as f:
+      dpickle.dump(title_pp, f)
+
+    # Save the processed data
+    np.save(data_dir + 'train_title_vecs.npy', train_title_vecs)
+    np.save(data_dir + 'train_body_vecs.npy', train_body_vecs)
+  else:
+    time.sleep(120)
+
+  # TODO(jlewi): Why do we need to block waiting for the file?
+  # I think this is because only the master produces the npy
+  # files so the other workers need to wait for the files to arrive.
+  # It might be better to make preprocessing a separate job.
+  while True:
+    if os.path.isfile(data_dir + 'train_body_vecs.npy'):
+      break
+    print("Waiting for dataset")
+    time.sleep(2)
+  encoder_input_data, doc_length = load_encoder_inputs(data_dir + 'train_body_vecs.npy')
+  decoder_input_data, decoder_target_data = load_decoder_inputs(data_dir + 'train_title_vecs.npy')
 
-while True:
-  if os.path.isfile(data_dir + 'train_body_vecs.npy'):
-    break
-  print("Waiting for dataset")
-  time.sleep(2)
-encoder_input_data, doc_length = load_encoder_inputs(data_dir + 'train_body_vecs.npy')
-decoder_input_data, decoder_target_data = load_decoder_inputs(data_dir + 'train_title_vecs.npy')
+  num_encoder_tokens, body_pp = load_text_processor(data_dir + 'body_pp.dpkl')
+  num_decoder_tokens, title_pp = load_text_processor(data_dir + 'title_pp.dpkl')
 
-num_encoder_tokens, body_pp = load_text_processor(data_dir + 'body_pp.dpkl')
-num_decoder_tokens, title_pp = load_text_processor(data_dir + 'title_pp.dpkl')
+  #arbitrarly set latent dimension for embedding and hidden units
+  latent_dim = 300
 
-#arbitrarly set latent dimension for embedding and hidden units
-latent_dim = 300
+  ##### Define Model Architecture ######
 
-##### Define Model Architecture ######
+  ########################
+  #### Encoder Model ####
+  encoder_inputs = tf.keras.layers.Input(shape=(doc_length,), name='Encoder-Input')
 
-########################
-#### Encoder Model ####
-encoder_inputs = tf.keras.layers.Input(shape=(doc_length,), name='Encoder-Input')
+  # Word embeding for encoder (ex: Issue Body)
+  x = tf.keras.layers.Embedding(
+        num_encoder_tokens, latent_dim, name='Body-Word-Embedding', mask_zero=False)(encoder_inputs)
+  x = tf.keras.layers.BatchNormalization(name='Encoder-Batchnorm-1')(x)
```

Comment:
@inc0 @hamelsmu It doesn't look like the output of BatchNormalization is used anywhere? Is this expected?
Does Keras just automagically connect the current layer to the last layer?

Reply:
@jlewi look at line 160 Below, it that the output of the batch norm layer is used as input to the GRU layer

Reply:
Ack. Resolved.